packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert' show json;
+import 'package:file/file.dart';
+import 'package:file/local.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart';
+import 'package:shared_preferences_platform_interface/shared_preferences_platform_interface.dart';
+
+import 'src/win32.dart' as win32;
+
+/// Wrapper to the win32 ffi functions for testing.
+class Win32Wrapper {
+  /// Calls the win32 function for [getLocalDataPath].
+  String getLocalDataPath() {
+    return win32.getLocalDataPath();
+  }
+
+  /// Calls the win32 function for [getModuleFileName].
+  String getModuleFileName() {
+    return win32.getModuleFileName();
+  }
+}
+
+/// The Windows implementation of [SharedPreferencesStorePlatform].
+///
+/// This class implements the `package:shared_preferences` functionality for the web.
+class SharedPreferencesWindows extends SharedPreferencesStorePlatform {
+  /// The name of the parent directory for [fileName].
+  final String fileDirectory = 'Flutter';
+
+  /// File system used to store to disk. Exposed for testing only.
+  @visibleForTesting
+  FileSystem fs = LocalFileSystem();
+
+  /// Wrapper for win32 ffi calls.
+  @visibleForTesting
+  Win32Wrapper win32Wrapper = Win32Wrapper();
+
+  String _localDataFilePath;
+  /// The path to the file in disk were the preferences are stored.
+  @visibleForTesting
+  String get getLocalDataFilePath {
+    if (_localDataFilePath != null) {
+      return _localDataFilePath;
+    }
+    _localDataFilePath = win32Wrapper.getLocalDataPath();
+    String appName = _getAppName();
+    // Append app-specific identifiers.
+    _localDataFilePath += '\\$fileDirectory\\$appName.ini';
+    return _localDataFilePath;
+  }
+
+  Map<String, Object> _cachedPreferences;
+  /// The in-memory representation of the map saved to a file in disk;
+  @visibleForTesting
+  Map<String, Object> get getCachedPreferences {
+    if (_cachedPreferences == null) {
+      File localDataFile = fs.file(getLocalDataFilePath);
+      if (!localDataFile.existsSync()) {
```

Comment:
Why does reading create the file?

Reply:
There is a chance that the user calls `getAll()` before writing anything in which case it would throw an exception. As I understand it, the file should be created by an installer anyway to ensure that it will be removed once the app is uninstalled. Do you see a potential problem with this?

Reply:
As we discussed, the installer will almost always *not* create the file, although it may create the folder where the file will go. Further, the developer is *much* more likely to use reasonable defaults for missing values when writing the code because they won't be using and installer. Further, it's just as likely that the app doesn't even have an installer. We should *not* assume there will be an installer.

Reply:
> in which case it would throw an exception

Only if you try to read it. For instance, this won't throw:

```
_cachedPreferences = {};
if (localDataFile.existsSync()) {
  String stringMap = localDataFile.readAsStringSync();
  if (!stringMap.isEmpty) {
    _cachedPreferences = json.decode(stringMap) as Map<String, Object>;
  }
}
return _cachedPreferences;
...
```

Reply:
I just tried that, but calling `readAsStringSync` with no file throws a `OSError: The system cannot find the file specified`.

Reply:
The code I suggested only reads the file when it's there.

Reply:
Maybe it wasn't clear which parts I was suggesting you replace ; here's the entire suggested function:
```
  Map<String, Object> get getCachedPreferences {
    if (_cachedPreferences == null) {
      _cachedPreferences = {};
      File localDataFile = fs.file(getLocalDataFilePath);
      if (localDataFile.existsSync()) {
        String stringMap = localDataFile.readAsStringSync();
        if (!stringMap.isEmpty) {
          _cachedPreferences = json.decode(stringMap) as Map<String, Object>;
        }
      }
    }
    return _cachedPreferences;
  }
```